File: 001-500/0340. Longest Substring with At Most K Distinct Characters.java
// Medium
// Sliding Window, Hash Table
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/longest-substring-with-at-most-k-distinct-characters/

import java.util.*;
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        if (k == 0) return 0;
        Map<Character, Integer> map = new HashMap<>();
        int left = 0, maxLen = 0;
        for (int right = 0; right < s.length(); right++) {
            // update map
            char c = s.charAt(right);
            map.put(c, map.getOrDefault(c, 0) + 1);
            
            // check the case to shrink window size
            while (map.size() > k) {
                char leftChar = s.charAt(left);
                map.put(leftChar, map.getOrDefault(leftChar, 0) - 1);
                
                // need to remove the char with 0 time appear
                if (map.get(leftChar) == 0) {
                    map.remove(leftChar);
                }
                left++;
            }
            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }
}

File: 0001-0500/0340. Longest Substring with At Most K Distinct Characters.py
# Time:O(n), Space:O(k)

class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        if k == 0 or not s:
            return 0
        
        left = 0
        max_len = 0
        char_count = {}

        for right in range(len(s)):
            char_count[s[right]] = char_count.get(s[right], 0) + 1

            while len(char_count) > k:
                char_count[s[left]] -= 1
                if char_count[s[left]] == 0:
                    del char_count[s[left]]
                left += 1
            
            max_len = max(max_len, right - left + 1)
        
        return max_len
    